File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/instruction/InstructionClassVisitor.java
package net.yakclient.mixin.base.internal.instruction;

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.tree.ClassNode;

public class InstructionClassVisitor extends ClassVisitor {
    private final String targetMethod;
    private final InstructionInterceptor interceptor;

    public InstructionClassVisitor(InstructionInterceptor interceptor, String targetMethod) {
        super(net.yakclient.mixin.base.YakMixins.ASM_VERSION, new ClassNode());
        this.interceptor = interceptor;
        this.targetMethod = targetMethod;
    }

    public Instruction getInstructions() {
        return this.interceptor.intercept();
    }


    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        final MethodVisitor visitor = super.visitMethod(access, name, desc, signature, exceptions);
        if (visitor != null && targetMethod != null && ByteCodeUtils.descriptorsSame(targetMethod, name + desc) ) {
            if (!(this.interceptor instanceof MethodVisitor)) throw new IllegalArgumentException("InstructionInterceptor must inherit from the MethodVisitor");
            return (MethodVisitor) this.interceptor;
        }
        return visitor;
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/instruction/InstructionClassVisitor.kt
package net.yakclient.mixins.base.internal.instruction

import net.yakclient.mixins.base.internal.bytecode.ByteCodeUtils
import net.yakclient.mixins.base.internal.bytecode.ByteCodeUtils.descriptorsSame
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.tree.ClassNode

class InstructionClassVisitor(
    private val interceptor: InstructionInterceptor,
    private val targetMethod: String
) : ClassVisitor(ByteCodeUtils.ASM_VERSION, ClassNode()) {

    val instructions: Instruction
        get() = interceptor.intercept()

    override fun visitMethod(
        access: Int,
        name: String,
        desc: String,
        signature: String?,
        exceptions: Array<String>?
    ): MethodVisitor {
        val visitor = super.visitMethod(access, name, desc, signature, exceptions)
        if (visitor != null && descriptorsSame(this.targetMethod, name + desc)) {
            require(this.interceptor is MethodVisitor) { "InstructionInterceptor must inherit from the MethodVisitor" }
            return this.interceptor
        }
        return visitor
    }
}